File: LeetCode/remote/ugly_number_II.rb
require 'test/unit'
extend Test::Unit::Assertions

# https://leetcode.com/submissions/detail/37365039/
# 596 / 596 test cases passed.
#   Status: Accepted
#   Runtime: 296 ms
#       
#   Submitted: 0 minutes ago
#
def nth_ugly_number(n)
    return 1 if n <= 1
    i2, i3, i5 = 0, 0, 0
    values = [1]
    for i in 1..n
        values[i] = [values[i2] * 2, values[i3] * 3, values[i5] * 5].min
        i2 += 1 if values[i] == values[i2] * 2
        i3 += 1 if values[i] == values[i3] * 3
        i5 += 1 if values[i] == values[i5] * 5
    end

    return values[n - 1]
end

# https://leetcode.com/submissions/detail/37364868/
# Time limit exceeded
# Solution is fine (relatively) but really slow
def nth_ugly_number_2(n)
    return 1 if n <= 1
    hash = {1 => true}
    start = 1
    count = 1
    while count <= n
        while hash[start].nil?
            start += 1
        end

        hash[start * 2] = true
        hash[start * 3] = true
        hash[start * 5] = true
        count += 1
        start += 1
    end

    return start - 1
end

ugly_numbers = [1, 2, 3, 4, 5, 6, 8, 9, 10, 12, 15, 16, 18, 20, 24, 25, 27, 30, 32, 36, 40, 45, 48, 50, 54, 60, 64, 72, 75, 80, 81, 90, 96, 100]

0.upto(ugly_numbers.size - 1).each { |i|
    assert_equal(ugly_numbers[i], nth_ugly_number(i + 1))
    assert_equal(ugly_numbers[i], nth_ugly_number_2(i + 1))
}

File: LeetCode/remote/ugly_number_II.cs
//  https://leetcode.com/problems/ugly-number-ii/
//
//   Write a program to find the n-th ugly number.
//
//   Ugly numbers are positive numbers whose prime factors only include 2, 3, 5.
//   For example, 1, 2, 3, 4, 5, 6, 8, 9, 10, 12 is the sequence of the first 10 ugly numbers.
//
//   Note that 1 is typically treated as an ugly number. 

//  https://leetcode.com/submissions/detail/53584642/
//
//  Submission Details
//  596 / 596 test cases passed.
//      Status: Accepted
//      Runtime: 104 ms
//          
//          Submitted: 0 minutes ago

public class Solution {
    public int NthUglyNumber(int n) {
        var factors = new [] { 0, 0, 0 };
        if (n <= 1)
        {
            return 1;
        }
        
        var result = new int[n + 1];
        result[0] = 1;
        for (var i = 1; i <= n; i++)
        {
            result[i] = new [] { result[factors[0]] * 2, result[factors[1]] * 3, result[factors[2]] * 5 }.Min();
            if (result[factors[0]] * 2 == result[i])
            {
                factors[0]++;
            }
            if (result[factors[1]] * 3 == result[i])
            {
                factors[1]++;
            }
            if (result[factors[2]] * 5 == result[i])
            {
                factors[2]++;
            }
        }
        
        return result[n - 1];
    }
}
